Update CharacterBody2D tutorial, code and descriptions (#6614)
godotengine/godot-docs: tutorials/physics/using_character_body_2d.rst

    # using move_and_collide
    var collision = move_and_collide(velocity * delta)
    if collision:
        velocity = velocity.slide(collision.get_normal())

    # using move_and_slide
    move_and_slide()

 .. code-tab:: csharp

    // using MoveAndCollide

>     var collision = MoveAndCollide(velocity * (float)delta);

Reviewer: Is this `velocity` variable meant to be the `CharacterBody2D.Velocity` property? In which case it needs to be `Velocity` with uppercase `V`.

    using Godot;

    public partial class CBExample : CharacterBody2D
    {
        public int Speed = 300;

        public void GetInput()
        {
            Vector2 inputDir = Input.GetVector("ui_left", "ui_right", "ui_up", "ui_down");
            Velocity = inputDir * Speed;
        }

        public override void _PhysicsProcess(double delta)
        {
            GetInput();

>             MoveAndCollide(Velocity * delta);

Reviewer: Since Vector2 uses `float` and delta is a `double`, it needs to be converted to `float` so it can be multiplied.

```suggestion
            MoveAndCollide(Velocity * (float)delta);
```

Author: Dang it - I remembered to change `delta` to a `double`...

    var Bullet = preload("res://Bullet.tscn")
    var speed = 200

    func get_input():
        # Add these actions in Project Settings -> Input Map.
        var input_dir = Input.get_axis("backward", "forward")
        velocity = transform.x * input_dir * speed
        if Input.is_action_just_pressed("shoot"):
            shoot()

    func shoot():
        # "Muzzle" is a Marker2D placed at the barrel of the gun.
        var b = Bullet.instantiate()
        b.start($Muzzle.global_position, rotation)

>         get_tree.root.add_child(b)

Reviewer: This was a typo, right?

```suggestion
        get_tree().root.add_child(b)
```

Author: Absolutely. :)

    func _physics_process(delta):
        get_input()
        var dir = get_global_mouse_position() - global_position
        # Don't move if too close to the mouse pointer.
        if dir.length() > 5:
            rotation = dir.angle()
            move_and_slide()

 .. code-tab:: csharp

    using Godot;

    public partial class CBExample : CharacterBody2D
    {
        private PackedScene _bullet = (PackedScene)GD.Load("res://Bullet.tscn");
        public int Speed = 200;

        public void GetInput()
        {
            // Add these actions in Project Settings -> Input Map.
            float inputDir = Input.GetAxis("backward", "forward");
            Velocity = Transform.x * inputDir * Speed;

>             if (Input.IsActionPressed("mouse_click"))

Reviewer: It seems like the action is called `shoot` in the GDScript version.

    extends CharacterBody2D

    var speed = 750

>     func start(_position, _direction):

Reviewer: I'm not very familiar with GDScript naming conventions but I don't think method parameters are usually prefixed by `_`.

Author: It's common when you're shadowing existing parameters like `position`.

    public partial class Bullet : CharacterBody2D
    {
        public int Speed = 750;

>         public void Start(Vector2 _position, float _direction)

Reviewer: The .NET naming convention for method parameters is camelCase.

```suggestion
        public void Start(Vector2 position, float direction)
```

        {
            Rotation = _direction;
            Position = _position;

>             _velocity = new Vector2(speed, 0).Rotated(Rotation);

Reviewer: ```suggestion
            Velocity = new Vector2(speed, 0).Rotated(Rotation);
```

        }

        public override void _PhysicsProcess(double delta)
        {

>             var collision = MoveAndCollide(_velocity * delta);

Reviewer: ```suggestion
            var collision = MoveAndCollide(Velocity * (float)delta);
```

            if (collision != null)
            {

>                 _velocity = _velocity.Bounce(collision.GetNormal());

Reviewer: ```suggestion
                Velocity = Velocity.Bounce(collision.GetNormal());
```

>                 if (collision.GetCollider().HasMethod("Hit"))

Reviewer: What do you think about caching the result of `collision.GetCollider()` to avoid making 2 calls? I'm not sure how much we should focus on performance in tutorials.

Author: In a basic tutorial like this, I don't think it should be that much of a concern.

    extends CharacterBody2D

    var speed = 300.0
    var jump_speed = -400.0

    # Get the gravity from the project settings so you can sync with rigid body nodes.
    var gravity = ProjectSettings.get_setting("physics/2d/default_gravity")

    func _physics_process(delta):
        # Add the gravity.
        velocity.y += gravity * delta

        # Handle Jump.
        if Input.is_action_just_pressed("jump") and is_on_floor():
            velocity.y = jump_speed

        # Get the input direction.
        var direction = Input.get_axis("ui_left", "ui_right")
        velocity.x = direction * speed

        move_and_slide()

 .. code-tab:: csharp

    using Godot;

    public partial class CBExample : CharacterBody2D
    {
        public float Speed = 100.0f;
        public float JumpSpeed = -400.0f;

        // Get the gravity from the project settings so you can sync with rigid body nodes.

>         public float Gravity = ProjectSettings.GetSetting("physics/2d/default_gravity");

Reviewer: `ProjectSettings.GetSetting` returns a `Variant` so it needs to be converted to `float`, you can cast it or use the `AsSingle` method, we use the latter in the [`basic_movement.cs`](https://github.com/godotengine/godot/blob/cd3e03432af9a638e342b5b79f76ca4d72f2a1f4/modules/mono/editor/script_templates/CharacterBody2D/basic_movement.cs#L12) template.

```suggestion
        public float Gravity = ProjectSettings.GetSetting("physics/2d/default_gravity").AsSingle();
```

        public override void _PhysicsProcess(double delta)
        {
            Vector2 velocity = Velocity;

            // Add the gravity.

>             velocity.y += Gravity * delta;

Reviewer: Oh, sorry, I forgot the cast in my previous comment.

```suggestion
            velocity.y += Gravity * (float)delta;
```

    extends CharacterBody2D

    var speed = 300.0
    var jump_speed = -400.0

    # Get the gravity from the project settings so you can sync with rigid body nodes.
    var gravity = ProjectSettings.get_setting("physics/2d/default_gravity")

    func _physics_process(delta):
        # Add the gravity.
        velocity.y += gravity * delta

        # Handle Jump.
        if Input.is_action_just_pressed("jump") and is_on_floor():
            velocity.y = jump_speed

        # Get the input direction and handle the movement/deceleration.
        var direction = Input.get_axis("ui_left", "ui_right")
        velocity.x = direction * speed

        move_and_slide()

 .. code-tab:: csharp

    using Godot;

    public partial class CBExample : CharacterBody2D
    {

>         public float Speed = 100.0;

Reviewer: It seems the GDScript version uses `300.0` as the initial value. Also, can't assign a `double` literal to a `float` variable, need to use the `f` suffix (or remove the `.0` because `int` literals can be implicitly converted to `float`).

```suggestion
        public float Speed = 100.0f;
```

>         public float JumpSpeed = -400.0;

Reviewer: ```suggestion
        public float JumpSpeed = -400.0f;
```

>         public float Gravity = ProjectSettings.GetSetting("physics/2d/default_gravity");

Reviewer: Missing comment from the GDScript version.

```suggestion

        // Get the gravity from the project settings so you can sync with rigid body nodes.
        public float Gravity = ProjectSettings.GetSetting("physics/2d/default_gravity");
```

        public override void _PhysicsProcess(double delta)
        {

>             Velocity.y += Gravity * delta;
>             if (Input.IsActionJustPressed("ui_accept") && IsOnFloor())
>                 Velocity.y = JumpSpeed;
>             Vector2 direction = Input.GetAxis("ui_left", "ui_right");
>             Velocity.x = direction * Speed;
>             MoveAndSlide();

Reviewer: Missing comments from the GDScript version. Also, the action seems to be called `jump`. And remember that you can't set the field of a property directly, need to create a separate velocity variable to update X and Y individually, then assign it back to the `Velocity` property of the object.

```suggestion
            Vector2 velocity = Velocity;

            // Add the gravity.
            velocity.y += Gravity * delta;

            // Handle Jump.
            if (Input.IsActionJustPressed("jump") && IsOnFloor())
                velocity.y = JumpSpeed;

            // Get the input direction and handle the movement/deceleration.
            Vector2 direction = Input.GetAxis("ui_left", "ui_right");
            velocity.x = direction * Speed;

            Velocity = velocity;
            MoveAndSlide();
```

Author: This is what comes from not using C# regularly. Thanks for the fixes!
